File: Globus/Store.m
//
//  Store.m
//  Denner
//
//  Created by Yves Bannwart-Landert on 14.03.12.
//  Copyright (c) 2012 youngculture ag. All rights reserved.
//

#import "Store.h"
#import "GlobusController.h"
#import "User.h"


@implementation Store

@synthesize storeId;
@synthesize channelName;
@synthesize city;
@synthesize name;
@synthesize address;
@synthesize phone;
@synthesize fax;
@synthesize longitude;
@synthesize latitude;
@synthesize manager;
@synthesize email;
@synthesize zip;
@synthesize images;
@synthesize holidays;
@synthesize shopClosed;

@synthesize openingTimes;

@synthesize location, distance, fullAddress;


- (Store *)initWithDictionary:(NSDictionary *)theDictionary
{
    self = [super init];
    if (self)
    {
        NSObject *value;
		
        value = [NSNumber numberWithInt:[[theDictionary objectForKey:@"id"] integerValue]];
        if (value && value != 0)
		self.storeId = (NSNumber *)value;
        
        value = [theDictionary valueForKey:@"channelName"];
        if (value && value != [NSNull null])
		self.channelName = (NSString *)value;
        
        value = [theDictionary valueForKey:@"city"];
        if (value && value != [NSNull null])
		self.city = (NSString *)value;
        
        value = [theDictionary valueForKey:@"name"];
        if (value && value != [NSNull null])
		self.name = (NSString *)value;
        
        value = [theDictionary valueForKey:@"address"];
        if (value && value != [NSNull null])
		self.address = (NSString *)value;
        
        value = [theDictionary valueForKey:@"phone"];
        if (value && value != [NSNull null])
		self.phone = (NSString *)value;
		
		value = [theDictionary valueForKey:@"fax"];
        if (value && value != [NSNull null])
		self.fax = (NSString *)value;
		
		value = [theDictionary valueForKey:@"manager"];
        if (value && value != [NSNull null])
		self.manager = (NSDictionary *)value;
        
        value = [theDictionary valueForKey:@"email"];
        if (value && value != [NSNull null])
		self.email = (NSString *)value;
        
        value = [NSNumber numberWithInt:[[theDictionary objectForKey:@"zip"] integerValue]];
        if (value && value != 0)
		self.zip = (NSNumber *)value;
        
        // Location
        double latt = [[theDictionary valueForKey:@"latitude"] doubleValue];
        double longt = [[theDictionary valueForKey:@"longitude"] doubleValue];
        
        if (latt && latt > 0 && longt && longt > 0)
        {
            self.latitude = [NSNumber numberWithDouble:latt];
            self.longitude = [NSNumber numberWithDouble:longt];
            
            CLLocationCoordinate2D coordinate2D;
            coordinate2D.latitude = latt;
            coordinate2D.longitude = longt;
            
            if (coordinate2D.latitude > 0.0 && coordinate2D.longitude > 0.0)
			self.location = [[CLLocation alloc] initWithLatitude:coordinate2D.latitude longitude:coordinate2D.longitude];
        }
        
		//Images
        value = [theDictionary valueForKey:@"images"];
        if (value && value != [NSNull null]) {
            self.images = (NSMutableArray *)value;
        }
		
        // Openting Times
        value = [theDictionary valueForKey:@"openingTimes"];
        if (value && value != [NSNull null]) {
            self.openingTimes = (NSMutableArray *)value;
        }
		
		// holidays
        value = [theDictionary valueForKey:@"holidays"];
        if (value && value != [NSNull null]) {
            self.holidays = (NSMutableArray *)value;
        }
		
		value = [theDictionary valueForKey:@"shopClosed"];
        self.shopClosed = value && value != [NSNull null];
    }
    return self;
}

- (Store *)initWithStore:(Store *)theStore location:(CLLocation *)theLocation
{
    self = [super init];
    
	if (self)
	{
		self = theStore;
		
		if (theLocation)
		distance = [theStore.location distanceFromLocation:theLocation];
	}
	return self;
}

- (NSString *)fullAddress
{
    NSMutableString *fullAddressString = [[NSMutableString alloc] init];
    
    if (self.address)
	[fullAddressString appendFormat:@"%@\n ", self.address];
    if (self.zip)
	[fullAddressString appendFormat:@"%@ ", self.zip];
    if (self.city)
	[fullAddressString appendFormat:@"%@ ", self.city];
    
    return fullAddressString;
}

- (NSString *)managerName
{
	NSString *lang = NSLocalizedString(@"All.LanguageCode", @"");
	NSString *corLang = [[[GlobusController sharedInstance] loggedUser] language];
	
	NSString *managerName = self.manager[corLang];
	if (!managerName)
		managerName = self.manager[lang];
	
	return managerName;
}


#pragma mark - MKAnnotation protocol

- (NSString *)title
{
	return self.name ? self.name : self.channelName;
}

- (NSString *)subtitle
{
    NSMutableString *addressForSubtitle = [[NSMutableString alloc] init];
    
    if (self.address)
	[addressForSubtitle appendFormat:@"%@, ", self.address];
    if (self.zip)
	[addressForSubtitle appendFormat:@"%@ ", self.zip];
    if (self.city)
	[addressForSubtitle appendFormat:@"%@", self.city];
    
	return addressForSubtitle;
}

- (CLLocationCoordinate2D)coordinate
{
    CLLocationCoordinate2D coordinate2D;
    coordinate2D.latitude = [self.latitude doubleValue];
    coordinate2D.longitude = [self.longitude doubleValue];
    
    if (coordinate2D.latitude > 0.0 && coordinate2D.longitude > 0.0)
	self.location = [[CLLocation alloc] initWithLatitude:coordinate2D.latitude longitude:coordinate2D.longitude];
    
	return self.location.coordinate;
}


#pragma mark - NSCoding

- (void)encodeWithCoder:(NSCoder *)encoder
{
	[encoder encodeObject:storeId forKey:@"id"];
    [encoder encodeObject:channelName forKey:@"channelName"];
    [encoder encodeObject:city forKey:@"city"];
    [encoder encodeObject:name forKey:@"name"];
    [encoder encodeObject:address forKey:@"address"];
    [encoder encodeObject:phone forKey:@"phone"];
	[encoder encodeObject:fax forKey:@"fax"];
    [encoder encodeObject:longitude forKey:@"longitude"];
    [encoder encodeObject:latitude forKey:@"latitude"];
	[encoder encodeObject:manager forKey:@"manager"];
    [encoder encodeObject:email forKey:@"email"];
    [encoder encodeObject:zip forKey:@"zip"];
	[encoder encodeObject:images forKey:@"images"];
    [encoder encodeObject:openingTimes forKey:@"openingTimes"];
	[encoder encodeObject:holidays forKey:@"holidays"];
    [encoder encodeObject:location forKey:@"location"];
}

- (id)initWithCoder:(NSCoder *)decoder
{
	if ((self = [self init]))
	{
		self.storeId = [decoder decodeObjectForKey:@"id"];
        self.channelName = [decoder decodeObjectForKey:@"channelName"];
        self.city = [decoder decodeObjectForKey:@"city"];
        self.name = [decoder decodeObjectForKey:@"name"];
        self.address = [decoder decodeObjectForKey:@"address"];
        self.phone = [decoder decodeObjectForKey:@"phone"];
		self.fax = [decoder decodeObjectForKey:@"fax"];
        self.longitude = [decoder decodeObjectForKey:@"longitude"];
        self.latitude = [decoder decodeObjectForKey:@"latitude"];
		self.manager = [decoder decodeObjectForKey:@"manager"];
        self.email = [decoder decodeObjectForKey:@"email"];
        self.zip = [decoder decodeObjectForKey:@"zip"];
		self.images = [decoder decodeObjectForKey:@"images"];
        self.openingTimes = [decoder decodeObjectForKey:@"openingTimes"];
		self.holidays = [decoder decodeObjectForKey:@"holidays"];
        self.location = [decoder decodeObjectForKey:@"location"];
	}
	return self;
}
/*
 - (MultiRowCalloutCell *)calloutCell {
 return [MultiRowCalloutCell cellWithImage:[UIImage imageNamed:@"Pin.png"]
 title:name
 subtitle:address
 userData:[NSDictionary dictionaryWithObject:self forKey:@"store"]];
 }
 */
@end

File: Globus/Store.swift
//
//  Store.swift
//  Globus
//
//  Created by Igor Vojinovic on 9/13/16.
//  Copyright © 2016 appculture. All rights reserved.
//

import UIKit

class Store {

    var id = 0
    var name = ""
    var channelName = ""
    var address = ""
    var zip = 0
    var city = ""
    var phone = ""
    var fax = ""
    var longitude = 0.0
    var latitude = 0.0
    var email = ""
    var managerDE = ""
    var managerFR = ""
    var managerEN = ""
    var images = [ImageModel]()
    var shopClosed = ""
    var openingTimes = [DayModel]()
    var holidays = [DayModel]()
    
    init(with json: [String : AnyObject]) {
        
        guard
            let idString = json["id"],
            let nameString = json["name"],
            let channelNameString = json["channelName"],
            let addressString = json["address"],
            let zipString = json["zip"],
            let cityString = json["city"],
            let phoneString = json["phone"],
            let faxString = json["fax"],
            let longitudeVal = json["longitude"],
            let latitudeVal = json["latitude"],
            let emailString = json["email"],
            let managerDEString = json["manager"]?["de"],
            let managerFRString = json["manager"]?["fr"],
            let managerENString = json["manager"]?["en"],
            let imagesArray = json["images"],
            let shopClosedString = json["shopClosed"],
            let openingTimesArray = json["openingTimes"] as? [AnyObject],
            let openingTimesDictionary = openingTimesArray.first as? [String : AnyObject],
            let openingTimesDays = openingTimesDictionary["days"],
            let holidaysArray = json["holidays"]  as? [AnyObject],
            let holidaysDictionary = holidaysArray.first as? [String : AnyObject],
            let holidayDays = holidaysDictionary["days"]
            else {
                print("Error while reading model store")
                return
        }
        id = idString as! Int
        name = nameString as! String
        channelName = channelNameString as!String
        address = addressString as! String
        zip = zipString as! Int
        city = cityString as! String
        phone = phoneString as! String
        fax = faxString as! String
        longitude = longitudeVal as! Double
        latitude = latitudeVal as! Double
        email = emailString as! String
        managerDE = managerDEString as! String
        managerFR = managerFRString as! String
        managerEN = managerENString as! String
        images = ImageModel.initWithArray(imagesArray: imagesArray as! [[String : AnyObject]])
        shopClosed = shopClosedString as! String
        openingTimes = DayModel.initWithArray(daysArray: openingTimesDays as! [[String : AnyObject]])
        holidays = DayModel.initWithArray(daysArray: holidayDays as! [[String : AnyObject]])
    }
    
    static func initWithArray(storesArray: [[String : AnyObject]]) -> [Store] {
        var parsedArray = Array<Store>()
        for storeItem in storesArray {
            let tempStoreItem = Store(with: storeItem)
            parsedArray.append(tempStoreItem)
        }
        return parsedArray
    }
    
}
